src/main/java/io/cdap/plugin/gcp/spanner/connector/SpannerConnector.java

```diff
@@ -0,0 +1,276 @@
+    return getTableData(instance, database, table, sampleRequest.getLimit());
+  }
+
+  @Override
+  public void test(FailureCollector failureCollector) throws ValidationException {
+    String project = config.tryGetProject();
+    if (project == null) {
+      failureCollector
+        .addFailure("Could not detect Google Cloud project id from the environment.", "Please specify a project id.");
+    }
+    // if either project  cannot be loaded , no need to continue
+    if (!failureCollector.getValidationFailures().isEmpty()) {
+      return;
+    }
+
+    try (Spanner spanner = getSpanner()) {
+      spanner.getInstanceAdminClient().listInstances();
+    } catch (Exception e) {
+      failureCollector.addFailure(String.format("Could not connect to Spanner: %s", e.getMessage()),
+                                  "Please specify correct connection properties.");
+    }
+  }
+
+  @Override
+  public BrowseDetail browse(BrowseRequest browseRequest) throws IOException {
+    SpannerPath path = new SpannerPath(browseRequest.getPath());
+    try (Spanner spanner = getSpanner()) {
+      if (path.isRoot()) {
+        // browse project to list all instances
+        return listInstances(spanner, browseRequest.getLimit());
+      }
+      String instance = path.getInstance();
+      String database = path.getDatabase();
+      String table = path.getTable();
+      if (database == null) {
+        return listDatabases(spanner, instance, browseRequest.getLimit());
+      }
+      if (table == null) {
+        return listTables(spanner, instance, database, browseRequest.getLimit());
+      }
+      return getTableDetail(spanner, instance, database, table);
+    }
+  }
+
+  @Override
+  public ConnectorSpec generateSpec(ConnectorSpecRequest connectorSpecRequest) throws IOException {
+    SpannerPath path = new SpannerPath(connectorSpecRequest.getPath());
+    ConnectorSpec.Builder specBuilder = ConnectorSpec.builder();
+    Map<String, String> properties = new HashMap<>();
+    properties.put(SpannerSourceConfig.NAME_USE_CONNECTION, "true");
+    properties.put(SpannerSourceConfig.NAME_CONNECTION, connectorSpecRequest.getConnectionWithMacro());
+
+    String instanceName = path.getInstance();
+    if (instanceName != null) {
+      properties.put(SpannerSourceConfig.NAME_INSTANCE, instanceName);
+    }
+    String databaseName = path.getDatabase();
+    if (databaseName != null) {
+      properties.put(SpannerSourceConfig.NAME_DATABASE, databaseName);
+    }
+    String tableName = path.getTable();
+    if (tableName != null) {
+      properties.put(SpannerSourceConfig.NAME_TABLE, tableName);
+      Schema schema = getTableSchema(instanceName, databaseName, tableName);
+      specBuilder.setSchema(schema);
+    }
+    return specBuilder.addRelatedPlugin(new PluginSpec(SpannerSource.NAME, BatchSource.PLUGIN_TYPE, properties))
+      .build();
+  }
+
+  private Schema getTableSchema(String instance, String database, String table) throws IOException {
```

Comment:
there is similar logic in `SpannerSource` on getting schema, can we just extract that out? https://github.com/data-integrations/google-cloud/blob/develop/src/main/java/io/cdap/plugin/gcp/spanner/source/SpannerSource.java#L245-L273

Reply:
done